### ISSUE TITLE
EAGAIN: waitpid on JDK11

### ISSUE THREAD

**Issue by fzakaria** [jruby/jruby#6477]
**Environment Information**
```
❯ jruby -v
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 11.0.8+0-adhoc..jdk11u-jdk-11.0.810 on 11.0.8+0-adhoc..jdk11u-jdk-11.0.810 +jit [linux-x86_64]
❯ uname -a
Linux fmzakari-glaptop 5.7.17-1rodete4-amd64 #1 SMP Debian 5.7.17-1rodete4 (2020-10-01) x86_64 GNU/Linux
❯ java -version
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment (build 11.0.8+0-adhoc..jdk11u-jdk-11.0.810)
OpenJDK 64-Bit Server VM (build 11.0.8+0-adhoc..jdk11u-jdk-11.0.810, mixed mode)
```

**Expected Behavior**

When the **--dev** flag is provided to JRuby it should work as normal.
The following works without the flag.

```bash
❯ jruby -e 'system "date"'
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib -Djruby.backtrace.style=full
Tue Nov 24 04:03:50 AM UTC 2020
```

**Actual Behavior**
When the flag is provided, the EAGAIN is raised.
```
❯ jruby --dev -e 'system "date"'
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib -Djruby.backtrace.style=full
Tue Nov 24 04:03:31 AM UTC 2020
Errno::EAGAIN: Resource temporarily unavailable - Resource temporarily unavailable
       getStackTrace at java/lang/Thread.java:1607
    getBacktraceData at org/jruby/runtime/backtrace/TraceType.java:242
        getBacktrace at org/jruby/runtime/backtrace/TraceType.java:53
    captureBacktrace at org/jruby/RubyException.java:371
            preRaise at org/jruby/exceptions/RaiseException.java:115
              <init> at org/jruby/exceptions/RaiseException.java:65
              <init> at org/jruby/exceptions/Exception.java:38
              <init> at org/jruby/exceptions/StandardError.java:38
              <init> at org/jruby/exceptions/SystemCallError.java:38
  constructThrowable at org/jruby/RubySystemCallError.java:133
         toThrowable at org/jruby/RubyException.java:349
                from at org/jruby/exceptions/RaiseException.java:81
   newRaiseException at org/jruby/Ruby.java:4125
     newErrnoFromInt at org/jruby/Ruby.java:3746
     newErrnoFromInt at org/jruby/Ruby.java:3765
     raiseErrnoIfSet at org/jruby/RubyProcess.java:1698
             waitpid at org/jruby/RubyProcess.java:919
              system at org/jruby/RubyKernel.java:1691
                call at org/jruby/RubyKernel$INVOKER$s$0$0$system.gen:-1
                call at org/jruby/internal/runtime/methods/JavaMethod.java:815
                call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
        cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:375
                call at org/jruby/runtime/callsite/CachingCallSite.java:174
         processCall at org/jruby/ir/interpreter/InterpreterEngine.java:316
           interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:72
      INTERPRET_ROOT at org/jruby/ir/interpreter/Interpreter.java:96
              <main> at -e:1
             execute at org/jruby/ir/interpreter/Interpreter.java:81
             execute at org/jruby/ir/interpreter/Interpreter.java:30
             execute at org/jruby/ir/IRTranslator.java:42
      runInterpreter at org/jruby/Ruby.java:1240
      runInterpreter at org/jruby/Ruby.java:1244
         runNormally at org/jruby/Ruby.java:1133
         runNormally at org/jruby/Ruby.java:1146
         runFromMain at org/jruby/Ruby.java:958
       doRunFromMain at org/jruby/Main.java:412
         internalRun at org/jruby/Main.java:304
                 run at org/jruby/Main.java:234
                main at org/jruby/Main.java:206
```

When I execute the same reproducer on JDK8 the issue does not occur.
There is a delta introduced with JDK11

```bash
❯ jruby -v
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 25.265-bga on 1.8.0_265-ga +jit [linux-x86_64]

❯ jruby --dev -e 'system "date"'
Picked up JAVA_TOOL_OPTIONS: -Djava.library.path=/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib
Mon Nov 23 08:05:17 PM PST 2020
```

**Comment by headius** [comment]
Sorry for the holiday delay! I will spend some cycles today looking into this.

**Comment by headius** [comment]
We were unable to reproduce this on @enebo's Fedora environment so we either need a sample NixOS env or you can work with us on gathering some diagnostic info.

My first suspect would be that for some reason dev mode is not able to load the C library bindings we use for things like `waitpid`. Try `-Xnative.verbose`.

Passing `-w` and/or `-d` may give us more info as well.

**Comment by fzakaria** [comment]
Thanks for looking at this -- here is some helpful reproductions information

Here is the extra information you requested
```bash
❯ jruby -Xnative.verbose -w -d --dev -e 'system "date"'
Successfully loaded native POSIX impl.
Wed Dec  2 09:16:33 AM PST 2020
Errno::EAGAIN: Resource temporarily unavailable - Resource temporarily unavailable
  waitpid at org/jruby/RubyProcess.java:919
   system at org/jruby/RubyKernel.java:1691
   <main> at -e:1
```

Here is a **reproducible** environment with Nix

```nix
let
  nixpkgs = import (builtins.fetchTarball {
    name = "nixos-unstable-2020-11-30";
    url =
      "https://github.com/nixos/nixpkgs/archive/24eb3f87fc610f18de7076aee7c5a84ac5591e3e.tar.gz";
    sha256 = "1ca14hhbinnz1ylbqnhwinjdbm6nn859j4gmyamg2kr7jl6611s0";
  }) { };
in with nixpkgs;
with stdenv;
with stdenv.lib;
mkShell {
  name = "jruby-eagain-shell";
  buildInputs = [
    ((jruby.override { jre = jdk11; }).overrideAttrs (oldAtrrs: rec {
      version = "9.2.13.0";
      src = fetchurl {
        url =
          "https://s3.amazonaws.com/jruby.org/downloads/${version}/jruby-bin-${version}.tar.gz";
        sha256 =
          "73a8c241a162e644c87e864c3485c55adedeb82a6fd80fa3cb538fdacda7af58";
      };
    }))
  ];
  shellHook = "";
}
```

1. Place the contents above in _shell.nix_ in a directory
2. Install Nix on any Linux distro `sh <(curl -L https://nixos.org/nix/install) --no-daemon`
3. Type `nix-shell` in the directory created in (1)
4. type `jruby -Xnative.verbose -w -d --dev -e 'system "date"'`

**Comment by fzakaria** [comment]
@enebo I have a nice Dockerfile for you that works with the above _shell.nix_ file

```
FROM nixos/nix

COPY shell.nix shell.nix

RUN nix-shell shell.nix --command "jruby -Xnative.verbose -w -d --dev -e 'system \"date\"'"
```

**Comment by fzakaria** [comment]
Just following up if my reproduction docker file was good enough. 🙏

**Comment by headius** [comment]
@fzakaria I'm sure it is good but we were not able to look at this in time for 9.2.14.0. I still assume this is an issue specific to your environment causing it to fail to find the right functions to call, similar to how you had to hardcode the path to libc.

**Comment by fzakaria** [comment]
Makes sense. 
Just had some sticky notes on things I wanted to follow-up on. I'll clear this one if it's good enough for reproduction.
Thanks! :)

**Comment by headius** [comment]
@fzakaria I have now reproduced locally using your Dockerfile and shell.nix!

**Comment by headius** [comment]
@fzakaria Looking at the options we pass to the JVM for `--dev` the most likely candidate is this property for disabling the generation of a native assembly stub:

https://github.com/jruby/jruby/blob/794f77bb111036364ba1f0fb26ba028df8c1ab01/bin/.dev_mode.java_opts#L5

Removing this option from the `.dev_mode.java_opts` file may be a workaround, but it is also possible that this was fixed by #6481 (released in 9.2.14.0). One of the many fixes that happened to the JNR subsystem over the summer was making the non-assembly logic work more like the assembly-based logic.

I will try to confirm this today, but if you get to it first let me know.

**Comment by headius** [comment]
Hmm... the ASM property does not break this, but `-Xcompile.mode=OFF` does. So something using this property follows a different path when calling `waitpid` logic.

**Comment by headius** [comment]
This is still an issue with how JNR is used from JRuby in dev mode. The `compile.mode=OFF` setting appears to disable the use of code-generated Java stubs for jnr-ffi targets, similar to how the asm flag above turns off the native stub. When the Java stubs are generated, we fall back on a generic invoker: https://github.com/jruby/jruby/blob/390ca2c47ebec80b0b01221c19b0ee23f75253d8/core/src/main/java/org/jruby/ext/ffi/jffi/BufferNativeInvoker.java

I suspect the problem is in this code.

Here is where we decide not to use the generated stubs:

https://github.com/jruby/jruby/blob/577d13a46bb08c8357fdab1ab12b0a35680402ab/core/src/main/java/org/jruby/ext/ffi/jffi/DefaultMethod.java#L83-L98

The `compilationFailed` value checked here is forced to true when the `compile.mode=false` property is set.

**Comment by headius** [comment]
This may be showing up in a JDK14 build intermittently. Here is one such example from Travis:

https://travis-ci.com/github/jruby/jruby/jobs/457507809#L1146

**Comment by headius** [comment]
I may have been mistaken about why the compile.mode value affects this. We use that fallback invoker for Ruby **ffi**, not for the internal FFI we use for binding `waitpid`. I am playing around with JRuby inside a container based on @fzakaria image today.

**Comment by fzakaria** [comment]
@headius sorry I haven't had a chance to play with it. Did you figure out how to enter the nix-shell in a tty ?
You can change the Dockerfile or docker command to override it and set it to `/bin/bash`; afterwards simply type `nix-shell` and it will enter you in a bash TTY with the correct JRuby

**Comment by headius** [comment]
@fzakaria Yes I was able to get a bash shell into the image you provided. Here is what I have learned:

* Two properties set by `--dev` are sufficient and required to cause this behavior: `jruby.compile.mode=OFF` and `jnr.ffi.asm.enabled=false`. The first turns off JRuby JIT and also modifies our Ruby FFI logic to not generate optimized JVM bytecode for FFI functions (instead using a generic fallback invoker). The second turns off generation of native assembly stubs on the native side of any jnr-ffi bound function, including those used by Ruby FFI.
* Calling `system` is equivalent to calling `spawn` followed by `Process.waitpid` and both forms exhibit the same behavior.
* Calling our jnr-posix `waitpid` directly, as shown below, does not appear to show the same problem.

Example calling `waitpid` via jnr-posix and checking errno, rather than calling `Process.waitpid` which does both:
```
p JRuby.runtime.posix.errno # =>0
pid = spawn 'date'
p JRuby.runtime.posix.errno # => 2
JRuby.runtime.posix.errno 0 # => clear errno to be sure, not necessary
p JRuby.runtime.posix.errno # => 0
JRuby.runtime.posix.waitpid(pid, nil, 0)
p JRuby.runtime.posix.errno # => 0
```

When run with the offending properties set, this shows that the errno after waitpid is zero and there is no error. If we instead call `Process.waitpid` it apparently sees an `EAGAIN` errno and raises.

I need to try to step through this logic and make sure the `Process.waitpid` logic is actually doing what I think it is.

Other notes:

* Disabling RubyGems with `--disable-gems` appears to prevent the problem. We suspect some FFI code is called during RubyGems boot that contributes to this behavior somehow.
* Disabling logic to use `pthread_kill` to interrupt a hung `waitpid` call does not appear to have an effect, but step debugging will confirm whether it is being successfully disabled.

**Comment by fzakaria** [comment]
Wow great follow-up; thank you for the deep dive.

**Comment by headius** [comment]
I am starting to suspect this is a false error.

The logic in `Process.waitpid` seems to be bad form in that it does not check the return value to know if there is an error.

In this case, `waitpid` returns the provided pid, indicating success. Instead of checking for a -1 return value, the code proceeds to check `errno` and if that is nonzero it assumes there was an error.

This is a safe assumption if the function is sanitary and makes sure to clear errno when there's no error. However, this may not be the case, either because of logic in `waitpid` that leaves an errno set, or because of logic in the unoptimized logic for the `waitpid` call making additional calls that corrupt the errno.

My current theory is that **there is no error**. 🥄

The logic should be modified to actually check the return value for an error state, at the very least. That would avoid the error in this case, since it seems to be a successful call to `waitpid`. The additional work to set all our minds at ease would be to determine *why* the errno appears to be set to `EAGAIN` after a successful `waitpid`.

Additional discoveries:

* Step debugging shows that errno is 11 (EAGAIN) immediately after our jnr-ffi `waitpid` call completes. None of the surrounding code in JRuby appears to be affecting it.
* An associate of mine @mrnoname1000 was able to reproduce on his Linux environment that does not involve Nix. His glibc version is 2.30, if that turns out to be important.

**Comment by headius** [comment]
I have pushed #6489 that fixes `waitpid` and `wait` to check for a -1 return value before raising errno.

**Comment by headius** [comment]
A few more discoveries:

* @mrnoname1000 was able to reproduce in an environment running with musl libc.
* Looping over the `spawn`+`waitpid` combination shows that this only fails the first time. Subsequent iterations do not leave a dangling EAGAIN and errno is zero as expected. I stepped into code all the way down to the native call and did not see any difference. There may still be some logic that is only running the first time on the native side, but that will require more investigation in the lowest laters of jffi.
* This has been seen across a range of Linux versions from 5.4 to 5.9. @enebo was unable to reproduce on Fedora with Linux 5.8.

I have pinged @therealaph for input since he recently contacted me about a similar issue (that turned out not to be JRuby's fault).
